server/src/test/java/org/elasticsearch/index/mapper/blockloader/GeoPointFieldBlockLoaderTests.java

```diff
@@ -0,0 +1,207 @@
+import org.elasticsearch.common.geo.GeoPoint;
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.utils.WellKnownBinary;
+import org.elasticsearch.index.mapper.BlockLoaderTestCase;
+import org.elasticsearch.index.mapper.MappedFieldType;
+
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class GeoPointFieldBlockLoaderTests extends BlockLoaderTestCase {
+    public GeoPointFieldBlockLoaderTests(BlockLoaderTestCase.Params params) {
+        super("geo_point", params);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected Object expected(Map<String, Object> fieldMapping, Object value, TestContext testContext) {
+        var extractedFieldValues = (ExtractedFieldValues) value;
+        var values = extractedFieldValues.values();
+
+        var nullValue = switch (fieldMapping.get("null_value")) {
+            case String s -> convert(s, null);
+            case null -> null;
+            default -> throw new IllegalStateException("Unexpected null_value format");
+        };
+
+        if (params.preference() == MappedFieldType.FieldExtractPreference.DOC_VALUES && hasDocValues(fieldMapping, true)) {
+            if (values instanceof List<?> == false) {
+                var point = convert(values, nullValue);
+                return point != null ? point.getEncoded() : null;
+            }
+
+            var resultList = ((List<Object>) values).stream()
+                .map(v -> convert(v, nullValue))
+                .filter(Objects::nonNull)
+                .map(GeoPoint::getEncoded)
+                .sorted()
+                .toList();
+            return maybeFoldList(resultList);
+        }
+
+        if (params.syntheticSource() == false) {
+            return exactValuesFromSource(values, nullValue);
+        }
+
+        // Usually implementation of block loader from source adjusts values read from source
+        // so that they look the same as doc_values would (like reducing precision).
+        // geo_point does not do that and because of that we need to handle all these cases below.
+        // If we are reading from stored source or fallback synthetic source we get the same exact data as source.
+        // But if we are using "normal" synthetic source we get lesser precision data from doc_values.
+        // That is unless "synthetic_source_keep" forces fallback synthetic source again.
+
+        if (testContext.forceFallbackSyntheticSource()) {
+            return exactValuesFromSource(values, nullValue);
+        }
+
+        String syntheticSourceKeep = (String) fieldMapping.getOrDefault("synthetic_source_keep", "none");
+        if (syntheticSourceKeep.equals("all")) {
+            return exactValuesFromSource(values, nullValue);
+        }
+        if (syntheticSourceKeep.equals("arrays") && extractedFieldValues.documentHasObjectArrays()) {
```

Comment:
This is an interesting case and i suspect it is not intentional.

Reply:
Can you elaborate why this isn't intentional?